File: app/src/main/java/com/jpwolfso/soundboardqt/SoundboardTileConfig.java
package com.jpwolfso.soundboardqt;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatImageButton;
import androidx.documentfile.provider.DocumentFile;

import android.Manifest;
import android.content.ComponentName;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.media.MediaPlayer;
import android.media.MediaRecorder;
import android.net.Uri;
import android.net.rtp.AudioStream;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.os.FileUtils;
import android.provider.MediaStore;
import android.provider.Settings;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.channels.FileChannel;
import java.nio.file.CopyOption;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.nio.file.attribute.FileAttribute;
import java.nio.file.attribute.FileOwnerAttributeView;
import java.text.SimpleDateFormat;
import java.util.Date;

import static android.content.Intent.ACTION_CREATE_DOCUMENT;
import static android.content.Intent.ACTION_GET_CONTENT;
import static android.content.Intent.EXTRA_TITLE;

public class SoundboardTileConfig extends AppCompatActivity {

    AppCompatImageButton recordButton;
    TextView status;
    MediaRecorder mediaRecorder;
    Context context;
    EditText editText;
    Button saveButton;
    File tempfile;
    File file;
    SharedPreferences sharedPreferences;
    SharedPreferences.Editor editor;
    String buttonversion;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_config);
        recordButton = findViewById(R.id.imageButton);
        recordButton.setOnClickListener(startRecord);
        status = findViewById(R.id.status);
        context = this;
        editText = findViewById(R.id.editText);
        editText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.length() > 0) {
                    saveButton.setEnabled(true);
                } else {
                    saveButton.setEnabled(false);
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        saveButton = findViewById(R.id.saveButton);
        saveButton.setOnClickListener(saveRecord);

        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.RECORD_AUDIO},0);
        }

        ComponentName componentName = getIntent().getParcelableExtra(Intent.EXTRA_COMPONENT_NAME);
        buttonversion = componentName.getShortClassName().substring(15);

        setTitle("Soundboard Config - Button " + buttonversion);

        file = new File(getFilesDir(),"recording" + buttonversion);
        tempfile = new File(getFilesDir(),"temprecording" + buttonversion);

        sharedPreferences = getSharedPreferences("buttons",MODE_PRIVATE);
        editor = sharedPreferences.edit();

    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        switch (grantResults[0]) {
                case (PackageManager.PERMISSION_GRANTED): {
                    if (requestCode == 1) {
                        exportSound();
                    } else if (requestCode == 2) {
                        importSound();
                    }
                    break;
                }
                case (PackageManager.PERMISSION_DENIED): {
                    AlertDialog.Builder dialog = new AlertDialog.Builder(this)
                            .setNegativeButton("Exit", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    finish();
                                }
                            })
                            .setCancelable(false);
                    if (requestCode == 0) {
                        dialog.setMessage("Soundboard Quick Tile requires permission to record audio")
                                .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        requestPermissions(new String[]{Manifest.permission.RECORD_AUDIO}, 0);
                                    }
                                });
                    } else if (requestCode == 1) {
                        dialog.setMessage("Soundboard Quick Tile requires this permission to export this button sound to your device storage")
                                .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 1);
                                    }
                                });
                    } else if (requestCode == 2) {
                        dialog.setMessage("Soundboard Quick Tile requires this permission to import a sound from your device storage")
                                .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 2);
                                    }
                                });
                    }
                    dialog.create();
                    dialog.show();
                    break;
                }
        }
    }

    View.OnClickListener startRecord = new View.OnClickListener() {
        @Override
        public void onClick(View v) {

            if (file.exists()) {
                file.delete();
            }

            editor.remove("button" + buttonversion).commit();

            recordButton.setImageResource(R.drawable.ic_stop);
            status.setText("Recording... Click to end recording");
            recordButton.setOnClickListener(endRecord);

            mediaRecorder = new MediaRecorder();
            mediaRecorder.setAudioSource(MediaRecorder.AudioSource.MIC);
            mediaRecorder.setOutputFormat(MediaRecorder.OutputFormat.DEFAULT);
            mediaRecorder.setOutputFile(tempfile.getAbsolutePath());
            mediaRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.DEFAULT);
            try {
                mediaRecorder.prepare();
            } catch (IOException e) {
                e.printStackTrace();
            }

            mediaRecorder.start();

        }
    };

    View.OnClickListener endRecord = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            mediaRecorder.stop();
            mediaRecorder.reset();
            mediaRecorder.release();
            mediaRecorder = null;
            recordButton.setImageResource(R.drawable.ic_record);
            status.setText("Enter a button title and click the Save button to save the quick tile, or click above to record again");
            recordButton.setOnClickListener(startRecord);
            saveButton.setVisibility(View.VISIBLE);
            editText.setVisibility(View.VISIBLE);

        }
    };

    View.OnClickListener saveRecord = new View.OnClickListener() {
        @Override
        public void onClick(View v) {

            String buttonName = editText.getText().toString();

            if (buttonName != null) {
                editor.putString("button" + buttonversion, buttonName).commit();
            }

            tempfile.renameTo(file);
            Toast.makeText(context, "Button saved successfully", Toast.LENGTH_SHORT).show();

            finish();
        }
    };

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_overflow, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case (R.id.appinfo): {
                Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                intent.setData(Uri.parse("package:" + getPackageName()));
                startActivity(intent);
                break;
            } case (R.id.soundimport): {
                if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                    requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 2);
                } else {
                    importSound();
                }
                break;
            } case (R.id.soundexport): {
                if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                    requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 1);
                } else {
                    exportSound();
                }
                break;
            }
        }
        return super.onOptionsItemSelected(item);
    }

    public void exportSound() {
        new AlertDialog.Builder(context)
                .setTitle("Please read")
                .setMessage("On the following screen, please browse to the save location and verify the filename ends with the .m4a file extension.")
                .setPositiveButton("Proceed", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        Intent intent = new Intent(ACTION_CREATE_DOCUMENT);
                        intent.setType("audio/*");
                        intent.putExtra(EXTRA_TITLE, getSharedPreferences("buttons", MODE_PRIVATE).getString("button" + buttonversion, "Button " + buttonversion) + ".m4a");
                        startActivityForResult(intent, 1);
                    }
                }).create().show();
    }

    public void importSound() {

        new AlertDialog.Builder(context)
                .setTitle("Please read")
                .setMessage("On the following screen, please browse and open the file to import.")
                .setPositiveButton("Proceed", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        Intent intent = new Intent(ACTION_GET_CONTENT);
                        intent.setType("audio/*");
                        startActivityForResult(intent, 0);
                    }
                }).create().show();

    }

    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == RESULT_OK) {
            if (requestCode == 0) {

                try {
                    InputStream inputStream = getContentResolver().openInputStream(data.getData());
                    Files.copy(inputStream,tempfile.toPath(),StandardCopyOption.REPLACE_EXISTING);
                    inputStream.close();

                    recordButton.setVisibility(View.GONE);
                    editText.setVisibility(View.VISIBLE);
                    saveButton.setVisibility(View.VISIBLE);
                    status.setText("Enter a button title and click the Save button to finish importing the selected file");


                } catch (IOException e) {
                    e.printStackTrace();
                    Toast.makeText(context, "Failed to import button sound :(", Toast.LENGTH_SHORT).show();
                }

            } else if (requestCode == 1) {

                try {
                    OutputStream outputStream = getContentResolver().openOutputStream(data.getData());
                    Files.copy(file.toPath(), outputStream);
                    outputStream.close();
                    Toast.makeText(context, "Button sound exported successfully", Toast.LENGTH_SHORT).show();

                } catch (IOException e) {
                    e.printStackTrace();
                    Toast.makeText(context, "Failed to export button sound :(", Toast.LENGTH_SHORT).show();
                }

            }
        }
    }
}

File: app/src/main/java/com/jpwolfso/soundboardqt/SoundboardTileConfig.kt
package com.jpwolfso.soundboardqt

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.Editor
import android.content.pm.PackageManager
import android.media.MediaRecorder
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.text.Editable
import android.text.TextWatcher
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.RadioButton
import android.widget.RadioGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatButton
import androidx.appcompat.widget.AppCompatEditText
import androidx.appcompat.widget.AppCompatImageButton
import androidx.appcompat.widget.AppCompatTextView
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

class SoundboardTileConfig() : AppCompatActivity() {

    private lateinit var recordButton: AppCompatImageButton
    private lateinit var status: AppCompatTextView
    private var mediaRecorder: MediaRecorder? = null
    private lateinit var context: Context
    private lateinit var editText: AppCompatEditText
    lateinit var saveButton: AppCompatButton
    private lateinit var tempfile: File
    private lateinit var file: File
    private lateinit var sharedPreferences: SharedPreferences
    private lateinit var editor: Editor
    private lateinit var saveas:View

    @SuppressLint("CommitPrefEdits")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_config)

        //configure layout variables and listeners
        recordButton = findViewById(R.id.imageButton)
        recordButton.setOnClickListener {startRecord()}
        status = findViewById(R.id.status)
        context = this
        editText = findViewById(R.id.editText)
        editText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {

                saveButton.isEnabled = s.isNotEmpty()
            }

            override fun afterTextChanged(s: Editable) {}
        })
        saveButton = findViewById(R.id.saveButton)
        saveButton.setOnClickListener {saveRecord()}

        tempfile = File(filesDir, "temprecording")
        sharedPreferences = getSharedPreferences("buttons", Context.MODE_PRIVATE)
        editor = sharedPreferences.edit()

        if (sharedPreferences.getBoolean("firstrun",true)) {
            welcomeDialog()
        }

        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(arrayOf(Manifest.permission.RECORD_AUDIO), 0)
        }

        reinitDlg()
    }

    private fun welcomeDialog() = AlertDialog.Builder(context)
            .setTitle("Welcome!")
            .setMessage("To begin, please manually add the 4 sound buttons to your quick settings panel, then you can begin recording sounds!")
            .setNeutralButton(R.string.OK) { dialog, which -> editor.putBoolean("firstrun",false).commit()}
            .create().show()

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (grantResults[0]) {
            PackageManager.PERMISSION_GRANTED -> {
                if (requestCode == 1) {
                    exportDialog()
                } else if (requestCode == 2) {
                    importDialog()
                }
            }
            PackageManager.PERMISSION_DENIED -> {
                val dialog = AlertDialog.Builder(this)
                        .setNegativeButton(getString(R.string.exit)) { _, _ -> finish() }
                        .setCancelable(false)
                when (requestCode) {
                    0 -> {
                        dialog.setMessage(getString(R.string.message_permission_audio))
                                .setPositiveButton("OK") { _, _ -> requestPermissions(arrayOf(Manifest.permission.RECORD_AUDIO), 0) }
                    }
                    1 -> {
                        dialog.setMessage(getString(R.string.message_permission_storage_export))
                                .setPositiveButton("OK") { _, _ -> requestPermissions(arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 1) }
                    }
                    2 -> {
                        dialog.setMessage(getString(R.string.message_permission_storage_import))
                                .setPositiveButton("OK") { _, _ -> requestPermissions(arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 2) }
                    }
                }
                dialog.create()
                dialog.show()
            }
        }
    }

    private fun startRecord() {
        recordButton.setImageResource(R.drawable.ic_stop)
        status.text = getString(R.string.status_recording_in_progress)
        recordButton.setOnClickListener() {endRecord()}
        mediaRecorder = MediaRecorder()
        mediaRecorder!!.setAudioSource(MediaRecorder.AudioSource.MIC)
        mediaRecorder!!.setOutputFormat(MediaRecorder.OutputFormat.DEFAULT)
        mediaRecorder!!.setOutputFile(tempfile.absolutePath)
        mediaRecorder!!.setAudioEncoder(MediaRecorder.AudioEncoder.DEFAULT)
        try {
            mediaRecorder!!.prepare()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        mediaRecorder!!.start()
    }

    private fun endRecord () {
        mediaRecorder!!.stop()
        mediaRecorder!!.reset()
        mediaRecorder!!.release()
        mediaRecorder = null
        recordButton.setImageResource(R.drawable.ic_record)
        status.text = getString(R.string.status_recording_complete)
        recordButton.setOnClickListener() {startRecord()}
        saveButton.visibility = View.VISIBLE
        editText.visibility = View.VISIBLE
    }

    private fun reinitDlg() {
        saveas = layoutInflater.inflate(R.layout.dialog_saveas,null)
        saveas.findViewById<RadioButton>(R.id.radioButton1).text = sharedPreferences.getString("button1", R.string.sound_1.toString())
        saveas.findViewById<RadioButton>(R.id.radioButton2).text = sharedPreferences.getString("button2", R.string.sound_2.toString())
        saveas.findViewById<RadioButton>(R.id.radioButton3).text = sharedPreferences.getString("button3", R.string.sound_3.toString())
        saveas.findViewById<RadioButton>(R.id.radioButton4).text = sharedPreferences.getString("button4", R.string.sound_4.toString())
    }

    private fun saveRecord() = AlertDialog.Builder(context)
            .setTitle(getString(R.string.title_select_button_write))
            .setView(saveas)
            .setOnDismissListener {reinitDlg()}
            .setPositiveButton(getString(R.string.confirm)) { _, _->
                val buttonName = editText.text.toString()
                when (saveas.findViewById<RadioGroup>(R.id.radioGroup).checkedRadioButtonId) {
                    R.id.radioButton1 -> {
                        editor.putString("button1", buttonName).commit()
                        file = File(filesDir, "recording1")
                    }
                    R.id.radioButton2 -> {
                        editor.putString("button2", buttonName).commit()
                        file = File(filesDir, "recording2")
                    }
                    R.id.radioButton3 -> {
                        editor.putString("button3", buttonName).commit()
                        file = File(filesDir, "recording3")
                    }
                    R.id.radioButton4 -> {
                        editor.putString("button4", buttonName).commit()
                        file = File(filesDir, "recording4")
                    }
                }
                tempfile.renameTo(file)
                Toast.makeText(context, getString(R.string.toast_button_saved_success), Toast.LENGTH_SHORT).show()
                finish()
            }
            .create().show()

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_overflow, menu)

        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.appinfo -> {
                val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
                intent.data = Uri.parse("package:$packageName")
                startActivity(intent)
            }
            R.id.soundimport -> {
                if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                    requestPermissions(arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 2)
                } else {
                    importDialog()
                }
            }
            R.id.soundexport -> {
                if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                    requestPermissions(arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 1)
                } else {
                    exportDialog()
                }
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun exportDialog() = AlertDialog.Builder(context)
            .setTitle(getString(R.string.title_select_button_export))
            .setMessage(getString(R.string.message_export_help))
            .setView(saveas)
            .setOnDismissListener {reinitDlg()}
            .setPositiveButton(getString(R.string.proceed)) { _, _ ->
                when (saveas.findViewById<RadioGroup>(R.id.radioGroup).checkedRadioButtonId) {
                    R.id.radioButton1 -> { file = File(filesDir, "recording1") }
                    R.id.radioButton2 -> { file = File(filesDir, "recording2") }
                    R.id.radioButton3 -> { file = File(filesDir, "recording3") }
                    R.id.radioButton4 -> { file = File(filesDir, "recording4") }
                }
                val intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
                intent.type = "audio/*"
                startActivityForResult(intent, 1)
            }.create().show()

    private fun importDialog() = AlertDialog.Builder(context)
            .setTitle(getString(R.string.title_please_read))
            .setMessage(getString(R.string.message_help_import))
            .setPositiveButton(getString(R.string.proceed)) { _, _ ->
                val intent = Intent(Intent.ACTION_GET_CONTENT)
                intent.type = "audio/*"
                startActivityForResult(intent, 0)
            }.create().show()

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == 0) {
                try {
                    val inputStream = contentResolver.openInputStream(data!!.data!!)
                    if (inputStream!!.available() >= 1000000) { // SoundPool does not work with files greater than 1 MB
                        Toast.makeText(this,getString(R.string.toast_error_large_file),Toast.LENGTH_LONG).show()
                        return
                    }
                    FileOutputStream(tempfile).use { stream -> inputStream.copyTo(stream)}
                    inputStream.close()
                    recordButton.visibility = View.GONE
                    editText.visibility = View.VISIBLE
                    saveButton.visibility = View.VISIBLE
                    status.text = getString(R.string.status_import_recording)
                } catch (e: IOException) {
                    e.printStackTrace()
                    Toast.makeText(context, getString(R.string.toast_error_import_failed), Toast.LENGTH_SHORT).show()
                }
            } else if (requestCode == 1) {
                try {
                    val outputStream = contentResolver.openOutputStream(data!!.data!!)
                    FileInputStream(file).use { stream -> stream.copyTo(outputStream!!)}
                    outputStream!!.close()
                    Toast.makeText(context, getString(R.string.toast_export_success), Toast.LENGTH_SHORT).show()
                } catch (e: IOException) {
                    e.printStackTrace()
                    Toast.makeText(context, getString(R.string.toast_export_failed), Toast.LENGTH_SHORT).show()
                }
            }
        }
    }
}